src/basic/process-util.c

```diff
@@ -275,21 +276,59 @@ int get_process_cmdline(pid_t pid, size_t max_length, bool comm_fallback, char *
 }
 
 void rename_process(const char name[8]) {
+        static char *mm = NULL;
+        static size_t mm_size = 0;
+        size_t l;
+        char *old_mm = NULL;
+        size_t old_mm_size = 0;
+
+        /* This is a like a poor man's setproctitle(). It changes the comm field, argv[0], and also the glibc's
+         * internally used name of the process. For the first one a limit of 16 chars applies; to the second one in
+         * many cases one of 10 (i.e. length of "/sbin/init") — however if we have CAP_SYS_RESOURCES it is unbounded;
+         * to the third one 7 (i.e. the length of "systemd". If you pass a longer string it will likely be
+         * truncated. */
+
         assert(name);
 
-        /* This is a like a poor man's setproctitle(). It changes the
-         * comm field, argv[0], and also the glibc's internally used
-         * name of the process. For the first one a limit of 16 chars
-         * applies, to the second one usually one of 10 (i.e. length
-         * of "/sbin/init"), to the third one one of 7 (i.e. length of
-         * "systemd"). If you pass a longer string it will be
-         * truncated */
+        l = strlen(name);
 
+        /* First step, change the comm field. */
         (void) prctl(PR_SET_NAME, name);
 
+        /* Second step, change glibc's ID of the process name. */
         if (program_invocation_name)
                 strncpy(program_invocation_name, name, strlen(program_invocation_name));
 
+        /* Third step, allocate some memory we can place the new name in. To be usable for PR_SET_MM_ARG_START this
+         * needs to be memory suitable for a stack, hence we allocate it with MAP_GROWSDOWN. */
+        if (mm_size < l+1) {
```

Comment:
So, turns out on x86-x64 (as the most relevant arch), MAP_GROWSDOWN is actually not even needed, and maybe on other it's neither. I'll hence drop the flag now. This means on archs which really need this the prctl() will fail, and we'll go into the fallback path, but that's entirely safe.

Also, I'll change the code to always update saved_argv[0]. There's no reason not to do this if the prctl() worked. After all, internal code might still look there, who knows.